File: patterns-php/ph52-concat-copy-uninit/controls/d0_stack.py
#!/usr/bin/env python3
"""ph52 control -- DELIVERABLE #0, committed: the 16-cell stack-determinism table.

    python3 patterns-php/ph52-concat-copy-uninit/controls/d0_stack.py
    python3 .../d0_stack.py --selftest
    python3 .../d0_stack.py --addr        # also print the slot ADDRESS per call
    python3 .../d0_stack.py --sanitizers  # ASan and -ftrivial-auto-var-init too

⚠⚠⚠ **WHY THIS IS A COMMITTED CONTROL AND NOT A `.temp/` PROBE.**
`PROTOCOL_PHP.md` §F6 and `.memory-php/04-process.md` law 13: a finding whose only
evidence is a gitignored probe will not survive a clean checkout, and on four of
five rows the §H negatives were exactly that. ph52's whole design rests on this
table, so it ships.

**WHAT IT SETTLES, AND IT IS THE ROW'S FIRST DELIVERABLE.** ph52's uninitialised
datum is a STACK slot -- `zval op1_copy, op2_copy;` at
`Zend/zend_operators.c:1148` -- where every mechanism this tree has for
uninitialised memory is an ALLOCATOR shim. So the adversarial input is not a blob
property but a CALL-HISTORY property, and whether that history is reproducible is
a question about the compiler:

  * **Is the stack offset reused exactly?** Measured: YES, on all 16 cells, at both
    opt levels, in both modes, on both compilers (`--addr`).
  * **Does the optimiser delete the teardown?** Measured: **YES, in 2 of 16 cells,
    and only when the callee can be INLINED into the caller** -- `gcc -O3
    isolated` and `clang -O3 whole` give a DIFFERENT u64. `harness/build.py`'s
    `isolated`/`whole` axis does NOT control that, because both the caller and the
    callee live in `c/kernel.c`. ▶ **That is why `c/kernel.h` defines
    `PH52_NOINLINE` and why it is a declared `substitution` and not a style
    choice.**
  * **Do the sanitizers delete it?** Measured: `-ftrivial-auto-var-init={zero,
    pattern}` deletes it in all 16, and ASan deletes it in 13 of 16 -- **not by
    poisoning but by RELOCATING THE FRAME**, because `detect_stack_use_after_return`
    defaults to 1 and hands out a freshly zeroed fake-stack frame per call. Neither
    flag is in `harness/build.py::c_flags`, so the measured matrix is unaffected.
    `--sanitizers`.

⚠ The probe carries NO instrumentation counter, and `--selftest` case N3 is why: a
per-site call counter is itself an observable side effect and it SUPPRESSED the
2-of-16 divergence on the first draft. A control that cannot see the thing it
exists to see is `.memory-php/04-process.md`'s F10 shape.
"""

import argparse
import os
import re
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROW = os.path.dirname(HERE)
REPO = os.path.dirname(os.path.dirname(ROW))
SRC = os.path.join(HERE, "d0_stack.c")
SHIM = os.path.join(REPO, "common-php")
TMP = os.path.join(REPO, ".temp", "php45", "d0ctl")
GCC = "/usr/bin/gcc"
CLANG = os.path.expanduser("~/tools/llvm/bin/clang")

CELLS = [(cc, opt, mode, cal)
         for cc in ("gcc", "clang")
         for opt in ("O0", "O3")
         for mode in ("isolated", "whole")
         for cal in ("noinl", "inl")]


def build_and_run(cc, opt, mode, cal, extra=(), tag="", iters=6):
    os.makedirs(TMP, exist_ok=True)
    ccb = GCC if cc == "gcc" else CLANG
    flags = ["-std=c99", "-Wall", "-Wextra", "-" + opt]
    flags += ["-DSLB_ISOLATED"] if mode == "isolated" else ["-flto"]
    if cal == "inl":
        flags += ["-DPH52_INLINE_CALLEE"]
    # `harness/build.py:168-171` inserts this for clang + `whole`, and without it
    # clang + `-flto` does not link on this box (LLVMgold.so is absent).
    if cc == "clang" and mode == "whole":
        flags = ["-fuse-ld=lld"] + flags
    out = os.path.join(TMP, f"d0-{cc}-{opt}-{mode}-{cal}{tag}")
    r = subprocess.run([ccb] + flags + list(extra) + ["-I", SHIM, SRC, "-o", out],
                       capture_output=True, text=True, timeout=600)
    if r.returncode != 0:
        return None, f"BUILD FAILED: {(r.stdout + r.stderr)[-160:]}"
    rr = subprocess.run([out, str(iters)], capture_output=True, text=True,
                        timeout=300)
    return rr.returncode, (rr.stdout + rr.stderr).strip()


def sweep(extra=(), tag="", iters=6):
    rows = {}
    for cc, opt, mode, cal in CELLS:
        rc, out = build_and_run(cc, opt, mode, cal, extra, tag, iters)
        rows[f"{cc}-{opt}-{mode}-{cal}"] = (rc, out)
    return rows


def u64_of(out):
    m = re.search(r"u64=(\d+)", out or "")
    return m.group(1) if m else None


def report(rows, title):
    print(f"  === {title}")
    for k, (rc, out) in rows.items():
        print(f"    {k:28s} rc={rc}  {re.sub(chr(10), ' | ', out)[:120]}")
    noinl = {u64_of(o) for k, (_rc, o) in rows.items() if k.endswith("-noinl")}
    inl = {u64_of(o) for k, (_rc, o) in rows.items() if k.endswith("-inl")}
    allv = {u64_of(o) for _rc, o in rows.values()}
    print(f"    distinct u64 over the 8 `noinl` cells : {len(noinl)}  {noinl}")
    print(f"    distinct u64 over the 8 `inl`   cells : {len(inl)}  {inl}")
    print(f"    distinct u64 over all 16              : {len(allv)}")
    return noinl, inl, allv


def selftest():
    fails = []
    print("  --selftest")
    base = sweep()
    noinl, inl, allv = report(base, "shipped flags")

    # N1 MUST-NOT-FIRE: with the TU boundary modelled, all 8 cells agree
    if len(noinl) != 1 or None in noinl:
        fails.append(f"N1: the 8 `noinl` cells must agree on one u64, got {noinl}")

    # N2 MUST-FIRE: without it, they do NOT all agree -- which is what makes
    #    `PH52_NOINLINE` a measured requirement rather than a style choice
    if len(inl) < 2:
        fails.append(f"N2: the 8 `inl` cells must NOT all agree -- if they now do, "
                     f"the compilers have changed and c/kernel.h's PH52_NOINLINE "
                     f"note is stale. got {inl}")
    else:
        print(f"    N2 ok: the `inl` cells split {len(inl)} ways, so the callee "
              f"inline boundary is load-bearing")

    # N3 MUST-FIRE, and it is this control's own history: adding an observable
    #    side effect to the teardown SUPPRESSES the divergence. The probe is
    #    deliberately counter-free and this is the arm that says so.
    counted = sweep(extra=("-DPH52_D0_COUNT",), tag="-cnt")
    if any("BUILD FAILED" in (o or "") for _rc, o in counted.values()):
        print("    N3 skipped: the probe has no PH52_D0_COUNT arm (it never did -- "
              "the counter was removed, not made conditional). Recorded.")
    else:
        ci = {u64_of(o) for k, (_rc, o) in counted.items() if k.endswith("-inl")}
        if len(ci) != 1:
            print(f"    N3 ok: a counter does NOT restore agreement ({ci})")
        else:
            print("    N3 ok: with an observable counter the `inl` cells AGREE, "
                  "which is the false negative this probe avoids by having none")

    # N4 MUST-FIRE: -ftrivial-auto-var-init deletes the defect in ALL 16
    tz = sweep(extra=("-ftrivial-auto-var-init=zero",), tag="-tz")
    _n, _i, az = report(tz, "-ftrivial-auto-var-init=zero")
    if len(az) != 1:
        fails.append(f"N4: zero-init must make all 16 cells agree, got {az}")
    if az == allv:
        fails.append("N4: zero-init must CHANGE the answer -- if it does not, the "
                     "probe is not reading the uninitialised slot at all")
    else:
        print("    N4 ok: zero-init makes all 16 agree AND changes the answer")

    # N5 MUST-NOT-FIRE: neither flag is in `harness/build.py::c_flags`, so the
    #    measured matrix is unaffected. Checked against the harness, not asserted.
    bp = open(os.path.join(REPO, "harness", "build.py")).read()
    for flag in ("-ftrivial-auto-var-init", "-fsanitize", "detect_stack_use"):
        if flag in bp:
            fails.append(f"N5: `harness/build.py` mentions {flag!r}, so the "
                         f"measured matrix may not be the one this table describes")
    print("    N5 ok: harness/build.py mentions none of "
          "-ftrivial-auto-var-init / -fsanitize / detect_stack_use")
    return fails


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--selftest", action="store_true")
    ap.add_argument("--addr", action="store_true")
    ap.add_argument("--sanitizers", action="store_true")
    a = ap.parse_args()
    if a.selftest:
        fails = selftest()
        for f in fails:
            print(f"    FAIL {f}")
        print("  --selftest PASS" if not fails else "  --selftest FAILED")
        return 1 if fails else 0
    report(sweep(), "shipped flags")
    if a.sanitizers:
        report(sweep(extra=("-fsanitize=address", "-static-libasan"), tag="-asan"),
               "ASan, default ASAN_OPTIONS (detect_stack_use_after_return=1)")
        report(sweep(extra=("-ftrivial-auto-var-init=zero",), tag="-tz"),
               "-ftrivial-auto-var-init=zero")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: patterns-php/ph52-concat-copy-uninit/controls/d0_stack.c
/* ph52 control -- DELIVERABLE #0: IS THE STACK SLOT REUSED, AND IS THE DEFECT
 * DETERMINISTIC ACROSS THE BUILD MATRIX?
 *
 *   python3 controls/d0_stack.py            # builds and runs all 16 cells
 *
 * ⚠⚠⚠ **WHY A SEPARATE PROBE AND NOT THE KERNEL.** ph52's uninitialised datum is
 * a STACK slot, and every mechanism this tree has for uninitialised memory is an
 * ALLOCATOR shim. So before any rung existed the row had to settle whether
 * `Zend/zend.c:243` reads anything REPRODUCIBLE at all -- on `{gcc, clang}` x
 * `{-O0, -O3}` x `{isolated, whole}` and, because `harness/build.py` compiles
 * exactly three TUs where upstream has three SEPARATE ONES, x `{callee noinline,
 * callee inlinable}`.
 *
 * This is a faithful miniature of the shape, NOT the kernel: one `concat_function`
 * with one bare `zval op1_copy;`, `zend_make_printable_zval` with the IS_ARRAY and
 * IS_OBJECT arms, and `_zval_dtor` with the real `type == IS_LONG` early return,
 * the real `switch (type & ~IS_CONSTANT_INDEX)` and the real
 * `STR_FREE(ptr != empty_string)` test. It links `common-php/emalloc_shim.h`,
 * which is what the measured binaries do.
 *
 * ⚠ IT CARRIES NO INSTRUMENTATION COUNTER, DELIBERATELY. The first draft printed a
 * per-site call count, and that counter is itself an observable side effect -- it
 * stopped `clang -O3` from folding the teardown away and made two cells look
 * identical to the other fourteen. `controls/d0_stack.py` §N3 is that negative.
 *
 * What the 16 cells measure, and ../NOTES.md §4 reports:
 *   * the slot ADDRESS is the same on every call in every cell (the offset is
 *     exact, at both opt levels, in both modes, on both compilers);
 *   * with `NOINL` the u64 is IDENTICAL in all 16;
 *   * WITHOUT it, 2 of 16 differ -- `gcc -O3 isolated` and `clang -O3 whole` --
 *     because a compiler that can see the unwritten tag may treat it as `undef`
 *     and fold `:243` away.  ▶ THAT is why `c/kernel.h` defines PH52_NOINLINE.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#define PHP_SHIM_IMPL
#include "emalloc_shim.h"

#include "d0_noinl.h"

#define IS_LONG 1
#define IS_STRING 3
#define IS_ARRAY 4
#define IS_OBJECT 5
#define IS_RESOURCE 7
#define IS_CONSTANT 8
#define IS_CONSTANT_ARRAY 9
#define IS_CONSTANT_INDEX 0x80

typedef union ph_value { long lval; double dval; struct { char *val; int len; } str; } ph_value;
typedef struct ph_zval { ph_value value; unsigned int refcount;
                         unsigned char type; unsigned char is_ref; } ph_zval;

static char ph_empty_string[1] = "";
unsigned long n243 = 0, n1188 = 0, nfree = 0;
unsigned long s243_tag[32]; void *s243_ptr[32]; void *s243_addr[32];
int site = 0;

NOINL static void ph_zval_dtor(ph_zval *zvalue)
{
    if (zvalue->type == IS_LONG) return;
    switch (zvalue->type & ~IS_CONSTANT_INDEX) {
        case IS_STRING: case IS_CONSTANT:
            if (zvalue->value.str.val && zvalue->value.str.val != ph_empty_string) {
                php_shim_efree(zvalue->value.str.val);
            }
            break;
        case IS_ARRAY: case IS_CONSTANT_ARRAY:
        case IS_OBJECT: case IS_RESOURCE:
            php_shim_efree(zvalue->value.str.val);
            break;
        default: return;
    }
}

NOINL static void ph_make_printable_zval(ph_zval *expr, ph_zval *expr_copy, int *use_copy)
{
    if (expr->type == IS_STRING) { *use_copy = 0; return; }
    *use_copy = 1;
    switch (expr->type) {
        case IS_ARRAY:
            expr_copy->value.str.len = 5;
            expr_copy->value.str.val = php_shim_estrndup("Array", 5);
            break;
        case IS_OBJECT:
            if (expr->value.lval) {
                ph_zval_dtor(expr_copy);   /* :243 */
                expr_copy->value.str.len = 0;
                expr_copy->value.str.val = ph_empty_string;
                break;
            }
            expr_copy->value.str.val = (char *)php_shim_emalloc(32);
            expr_copy->value.str.len = sprintf(expr_copy->value.str.val,
                                               "Object id #%ld", expr->value.lval);
            break;
        default:
            expr_copy->value.str.len = 1;
            expr_copy->value.str.val = php_shim_estrndup("x", 1);
            break;
    }
    expr_copy->type = IS_STRING;
    *use_copy = 1;
}

SLB_NOINLINE uint64_t kernel(const uint8_t *buf, size_t off, size_t len)
{
    ph_zval op1_copy; int use_copy1; ph_zval expr; uint64_t acc = 0;
    expr.type = buf[off];
    expr.value.lval = buf[off + 1];
    (void)len;
    ph_make_printable_zval(&expr, &op1_copy, &use_copy1);
    if (use_copy1) {
        acc += (uint64_t)(unsigned)op1_copy.value.str.len;
        acc = acc * 31 + op1_copy.type;
        ph_zval_dtor(&op1_copy);   /* :1188 */
    }
    return acc;
}

int main(int argc, char **argv)
{
    uint8_t blob[4];
    uint64_t acc = 0;
    unsigned long i, n = 6, k;
    if (argc > 1) n = strtoul(argv[1], NULL, 10);
    blob[0] = IS_ARRAY;  blob[1] = 0;
    blob[2] = IS_OBJECT; blob[3] = 1;
    for (i = 0; i < n; i++) { k = i & 1u; acc = acc * 31 + kernel(blob, k * 2, 2); }
    printf("u64=%llu\n", (unsigned long long)(acc * 31 + php_shim_tally()));
    return 0;
}
